### ISSUE TITLE
Support overriding appsettings.json with appsettings.development.json generally

### ISSUE THREAD

**Issue by Skrypt** [OrchardCMS/OrchardCore#5808]
I just finished testing the OrchardCore.Azure.Shell module features and it works fine.
Though after review I thought that it would be a nice addition to be able to use also alternate settings while on dev but when using this feature to sync the dev settings with every developper.

The current implementation is fine for now as it won't allow to use such alternate settings file which is consistent. I've tested also using an alternate appsettings.development.json file on the root folder of the OrchardCore.Cms.Web project and it doesn't take it into consideration either. So it is consistent in the entire app. We can't use alternates.

**Comment by hishamco** [comment]
Is `appsettings.development.json` not supported yet?!!

**Comment by Skrypt** [comment]
Not concluant from my tests and I'm using latest dev.

**Comment by Skrypt** [comment]
Ok I made a test to see if it works and the appsettings.Development.json file in the OrchardCore.Cms.Web folder is not taken into consideration. We have code that loads it and we are aware that it exists in the code but seems like it's still taking appsettings.json instead of appsettings.Development.json.

**Comment by Skrypt** [comment]
Last test is concluant. You need to rename appsettings.json to appsettings.Development.json in OrchardCore.Cms.Web folder. And it is not case sensitive on Windows, need to see on Linux.

**Comment by hishamco** [comment]
> I think appsettings.Development.json might be case sensitive.

Yep, in Linux if I'm not wrong

**Comment by Skrypt** [comment]
I need to see why we don't just make the appsettings.Development.json with the higher priority over appsettings.json. Else we need to remove one file that could be kept there for eventual publishing purposes.

**Comment by jtkech** [comment]
@Skrypt 

As i retried it yesterday, appsettings.Development.json has an higher priority.

But i trust you ;)

So, when i will have time this night we could have a call on skipe to see why it doesn't work on your side.

**Comment by Skrypt** [comment]
Well it works but seems like the priority is higher on appsettings.json so if I keep that file in the folder then it will always take it even on Development environment.

**Comment by Skrypt** [comment]
Your Skype is not open since last night 😉

**Comment by jtkech** [comment]
Okay will open it later this night ;)

**Comment by Skrypt** [comment]
Conclusion is that it works. There could be things that we could do to make the service not load if the settings are not provided in the appsettings.Development.json file but for now I think it's fine how it is. I'm closing the issue.

**Comment by sebastienros** [comment]
Can you comment on the reason why you reopened it?

**Comment by Skrypt** [comment]
Because most services that we register from the DI doesn't take in consideration when there is no configurations set. Normally when you use a appsettings.development.json file it is for overriding values for development. 

Facts : 
- The tenants doesn't take it into consideration.
- The OrchardCore.Cms.Web project root appsettings.json file takes it into consideration.

So if I want for example to override the OrchardCore.Shells.Azure module settings with a appsettings.development.json file, **I can**, but it breaks the feature with throwing NRE's. After discussing with @deanmarcussen and @jtkech we ended up with different resolution conclusions.
I personally think that we should allow doing overrides as it is something standard with .NET Core.

**Comment by deanmarcussen** [comment]
> Because most services that we register from the DI doesn't take in consideration when there is no configurations set. Normally when you use a appsettings.development.json file it is for overriding values for development.

Yes, but the normal override is to provide a different value, not to provide a null value, and see the feature turn itself off. So for Azure, you might provide a `UseDevelopmentStorage=true` to use the development storage, but a blank value, would throw a null ref

> * The tenants doesn't take it into consideration.

Also true, mostly because they are mutable, and it is hard to know whether to write to the tenants `appsettings.json` or `appsettings.Development.json` (much room for confusion to occur here)

When I want to do this, for a single `Default` shell, I move all the configuration out of the tenants to the root `appsettings.json` and a `appsettings.Development.json` as it is a preconfigured tenant/shell

> * The OrchardCore.Cms.Web project root appsettings.json file takes it into consideration.
> 
> So if I want for example to override the OrchardCore.Shells.Azure module settings with a appsettings.development.json file, **I can**, but it breaks the feature with throwing NRE's. After discussing with @deanmarcussen and @jtkech we ended up with different resolution conclusions.
> I personally think that we should allow/support doing overrides as it is something standard with .NET Core.

Not sure about @jtkech conclusion, but this was my one for managing `Azure.Shells` in different environments
```
    public class Startup
    {
        private readonly IHostEnvironment _env;

        public Startup(IHostEnvironment env)
        {
            _env = env;
        }

        public void ConfigureServices(IServiceCollection services)
        {
            var builder = services.AddOrchardCms();
            if (!_env.IsDevelopment())
            {
                builder.AddDatabaseShellsConfiguration();
            }
        }
    }
```

or to use the Azure Storage Emulator, for local dev.

I am not sure that there is a good argument for needing to do this though (keeping tenant settings in sync with development), when we are talking about a large quantity of tenants, which the Azure.Shells / Database feature is intended for - it's not really intended for a site with a single shell

**Comment by jtkech** [comment]
Yes with @Skrypt we ended up to do the same thing.

The `appsettings.OtherEnv.json` doesn't replace `appsettings.json`, it is just added to the stack, and each time i tried it i could see that it is able to override the config values.

Note: As soon as a section is defined with one field of an options class, when we bind it to this options class, we get a non null value, even if an `appsettings.OtherEnv.json` set all the fields to null.

So on startup we could check e.g if `ConnectionString` is null, or having another property as e.g `Disabled` that `appsettings.OtherEnv.json` could override. Meanwhile someone could use its own derived options class with this property to be bound by the same config section.

But if we don't need to check the options but just need to check if this app level "functionality" is disabled or not, we could just use the additional config value that we need and under the same section for clarity, and then check it on startup.

Then this `Disabled` term could be a convention to disable an app level "functionality" so that we could automatically disable it without having to add code in the main startup, but as @deanmarcussen said we would need to pass to the `OrchardCoreBuilder` the already available `IConfiguration`.

I found a way to retrieve the main `IConfiguration` from the services collection, so we could add a related field to our `OrchardCoreBuilder` and set it when we build it, and use it in our extensions.

    var mainConfiguration = services
        .LastOrDefault(d => d.ServiceType == typeof(IConfiguration))
        ?.ImplementationFactory.Invoke(null);
        ;

**The above factory doesn't use the service provider parameter so we can pass a null value** [see here](https://github.com/dotnet/aspnetcore/blob/master/src/Hosting/Hosting/src/WebHostBuilder.cs#L276)

Meanwhile in `appsettings.AnyOtherEnv.json` we could have

    "OrchardCore": {
      "OrchardCore_Shells_Database": {
        "Disabled": true
      }

And then in the main startup

    public class Startup
    {
        private readonly IConfiguration _configuration;

        public Startup(IConfiguration configuration)
        {
            _configuration = configuration;
        }

        public void ConfigureServices(IServiceCollection services)
        {
            var builder = services.AddOrchardCms();

            var disabled = _configuration
                .GetSection("OrchardCore:OrchardCore_Shells_Database")
                .GetValue<bool>("Disabled");

            if (!disabled)
            {
                builder.AddDatabaseShellsConfiguration();
            }
        }
